### DISCUSSION TITLE
Profiling Godot Jolt and a PR for Godot

### DISCUSSION THREAD

**Discussion by jrouwe** [godot-jolt/godot-jolt#611] [General]
Hello,

Because [people keep making comparison videos between Godot Jolt and Unity](https://x.com/Saulcava1/status/1705896155249328206?s=20), and I'd like to win that battle 😄, I've done some profiling using Superluminal (great tool).

I've created a simple scene with 2000 boxes (which is what most people seem to do):

[StackOfBoxes.zip](https://github.com/godot-jolt/godot-jolt/files/12718860/StackOfBoxes.zip)

It has VSYNC turned off because that's what takes most time in the profiler and I was running it with the template_release build and run without editor. On my i9-12900H / 3070 Ti laptop this reaches up to 2030 fps (so I'm not sure why people only get 50 fps).

Draw takes most (57%), followed by the Jolt Physics update (24%), and as you already mentioned [here](https://x.com/thatmihe/status/1704850984735023419?s=20) the state synchronization takes a lot of time (14%). It calls a callback that goes through 5 (!) trampoline functions **per body**, but that's not what I wanted to talk about (yet). 

I noticed the following:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/89184edc-4582-417e-b0a5-834d4f40f896)

I.e. godot is calling ```_sync_body_state``` 2x per ```_body_state_changed``` callback, which is a waste because most of the time ```_integrate_forces``` is not registered.

So I created the following commit:

https://github.com/godotengine/godot/commit/eff7f27c48faaee4de915db598705d7946d97a0f

And that increases performance of the scene by 2.5% to up to 2080 fps on my machine.

Now before I create a PR and make an idiot of myself, what do you think about this PR?

**Comment by mihe** [comment]
I think that change looks great.

It is however a bit of a proactive one, since that second call to `_sync_body_state` was added [fairly recently](https://github.com/godotengine/godot/pull/79977) (and even looked over by yours truly 😬) so isn't part of Godot 4.1, which I assume is the version most people are comparing against. So you would help prevent a regression, rather than improve things when comparing to Godot 4.1, technically speaking.

(You'll want to make the same change for `PhysicalBone3D` as well as `RigidBody2D`, by the way.)

But yes, I've been seeing a few of these comparisons as well lately, and I'm not sure how to feel about them. It feels a bit "apples and oranges" to compare not only PhysX and Jolt, but also their integrations into two very different game engines. On top of that it's my impression that Godot as a project seems to not really lean into the typical min-maxing of performance at the cost of everything else that you find more traditional engines do, which I feel makes the comparison even more skewed.

... but that's of course not an argument against some easy wins like the improvement you found there.

I'm certainly curious about the overhead with the trampoline functions you mentioned. Last I checked Superluminal the majority of `JoltSpace3D::call_queries` seemed to consist of `Variant` stuff related to the crossing the extension boundaries back and forth multiple times. But I guess for something done as frequently as the state synchronization every bit of overhead adds up quickly.

Also, as an aside, keep in mind when you're working against Godot's `master` branch that Godot Jolt is compiled against the Godot 4.1 extension API (unless you went through [the trouble](https://github.com/godot-jolt/godot-jolt/blob/master/docs/hacking.md#updating-godot) of retargeting it). While they strive to not introduce any breaking changes these days, there is at least [one active breakage](https://github.com/godotengine/godot/issues/80444) related to physics that'll make Godot stomp memory when you do ray-casts using a 4.1 extension. So if you run into any weird crashes or behavior, that could be why.

**Reply by jrouwe** [reply]
Thanks for the tips! 

I created a PR [here](https://github.com/godotengine/godot/pull/82393).

> I'm certainly curious about the overhead with the trampoline functions you mentioned. 

I will post some more info on this later, I need to do some further analysis first.

**Comment by Saul2022** [comment]
Pretty good benchmark, though for context i test with an amd radeon vega 3 driver igpu,  the issue comes when stacking  more than 2k cubes  the framerrate doesn't seem to hold up as well, for example from 70 to 6  with 1.2k(the 70fps one) and then with 2.4k (the 6 fps one)  which will that can stutter with  physics as it can directly impact weaker systems like mobile.  It seem's that at some point the memory  falls up or there's a leak maybe, anyway will test 4.2 when jolt is avaliable there And i also tried changing solver  iterations,  decreasing , and doesn't seem to help. I say this for mobile gpus and the weaker ones, it obvious that on op systems it can handle over 20k potatoes xd.

**Reply by ywmaa** [reply]
Well 1K cube physics on an IGPU, that's still impressive for me as a 3D game Dev.

**Reply by Saul2022** [reply]
Right, it is, but for weaker device like mobile, the cost is bigger and may struggle, though curiously i seen other videos of engines that seem to have a similar issue on massive frame drops, like stride bullet physics, or i have the feel that flax may have it too, as i seen some tweets talking about it being slower than unity too despite being in a higher version. So mihe is right in that front.

**Reply by jrouwe** [reply]
@Saul2022 There are a couple of things you can try. First of all, I can see in your capture that you have (DEBUG) in your window, which means you run an ```editor``` build instead of a ```template_release``` build. On my machine this adds 20% more FPS.

Second of all, I'm curious what happens if you remove the MeshInstance3D from your RigidBody3D (only simulate the boxes, not render them). That would give an indication of where your bottleneck is: rendering on IGPU or simulating physics.

**Reply by Saul2022** [reply]
I don't think boxes mesh  are what takes performance as it basically just 4k primitives or something like that, ,why my igpu can handle without issues. 

As for debug unity also takes that build and ye it gives better performance. The issue seem's to be when a lot of bodies collide with each as if memory going out or being unused, which is weird  , because godot 4 uses DOD for servers.

And physics taking the most when i profilled, though looking at your profiling  may fix that.

**Comment by jrouwe** [comment]
Found [a small performance](https://github.com/jrouwe/JoltPhysics/pull/719) optimization in Body::GetInverseInertia which is called a ton from _sync_body_state.

There is potentially another small perf improvement here:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/304f0036-dcf0-4cd7-baba-fbf19b4769de)

These 2 loops can potentially be combined into 1 to avoid cache misses + extra calls to try_get.

The main improvement to godot that could be done is to change this function:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/e8b6c939-e8d9-4653-bd12-94c26fd0acd9)

so that it only calls the GDExtension 1x, which would allow you to get the body only once and re-use some calculations between the function calls:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/36241777-e573-424e-9a5f-04700afe87de)

In general I'm quite impressed with how little overhead godot-jolt adds :heart:

**Reply by mihe** [reply]
> These 2 loops can potentially be combined into 1 to avoid cache misses + extra calls to try_get.

This was actually the case a long time ago, before I even pushed this to GitHub. The problem (if I remember correctly) is that it changes the order in which `_integrate_forces` happens versus when the enter/exit signals are emitted from `Area3D`, which happen as part of its `call_queries`. Right now you're guaranteed to call all `_integrate_forces` before any `Area3D` signals are emitted. While that maybe doesn't feel like a terribly important thing to preserve compatibility with, it was nevertheless something that I wasn't (and still am not) super comfortable with breaking, despite the minor performance hit.

> The main improvement to godot that could be done is to change this function:

Yes, batching would certainly help. However, I think it's important to point out that `RigidBody3D` is just one "front-end" node for the `PhysicsServer3D::body_*` methods. It's entirely possible to create your own version of the `RigidBody3D` node that cares about a completely different set of states, so any sort of batching should ideally take that into account and not only cater to the set of states that `RigidBody3D` cares about. So you would likely need to bundle up all of the state that you can get from `PhysicsDirectBodyState3D` in some form. I'm not entirely sure what that would look like though.

With regards to `JoltBodyAccessor3D` showing up so much, I've been thinking about just dropping the whole body-locking thing wherever possible, since it currently doesn't do [any actual locking](https://github.com/godot-jolt/godot-jolt/blob/12f50371984a01dd8207b0b9fb347edff3de5d2a/src/spaces/jolt_space_3d.cpp#L265) in distributed builds anyway. Everything is also very sequential in Godot, and it's not like you're gonna have other threads ripping out bodies as you're trying to access them, especially not if you're using something like `PhysicsServer3DWrapMT`, so just storing the actual `JPH::Body*` should be fine I feel like. Similarly, the copying of bodies by using `PhysicsSystem::GetBodies` is not really necessary. All of this was mostly just me preparing for some future fine-grained locking that never actually happened.

> In general I'm quite impressed with how little overhead godot-jolt adds ❤️

I'm happy to hear that! While I haven't really profiled much (or at all really) during the development of this project, I've tried to be mindful of the typical performance pitfalls, which seems to have worked out fairly well. Some of it is also certainly thanks to Godot providing a (mostly) reasonable extension API as well.

**Reply by jrouwe** [reply]
> So you would likely need to bundle up all of the state that you can get from ```PhysicsDirectBodyState3D``` in some form. I'm not entirely sure what that would look like though.

My guess is that most of the users are going to go through ```RigidBody3D``` though, and those that aren't can still use the existing functionality. I would see it as a new function that delivers the exact state that ```RigidBody3D``` needs. 

Another option would be to remove  the ```inverse_inertia_tensor``` member from ```RigidBody3D``` and in the ```get_inverse_inertia_tensor``` do something similar to ```RigidBody3D::get_contact_count``` and fetch the value on the fly. I think it is unlikely that this will be called frequently, so it's better to pay the cost when someone is interested instead of for every body, every frame.

> So just storing the actual JPH::Body* should be fine I feel like. Similarly, the copying of bodies by using PhysicsSystem::GetBodies is not really necessary. All of this was mostly just me preparing for some future fine-grained locking that never actually happened.

I think this is a good idea.

**Comment by Saul2022** [comment]
I tried profiling with my test scene of 2.4k cubes, though don’t have much time , because they don’t let me use the  computer, definetly the bottlencek is not rendering , but physics which take  based on initial profiling 60ms or  80ms in total, though don’t trust a lot the numbers, because i didn’t have time to check or make screenshots from it. What i am curious what profiler to use to calculate memory usage?, Because as far as i know jolt still says that it lacks that on  the page.  I know it not most helpful, but i think the bottleneck comes with memory use.

**Reply by Saul2022** [reply]
https://github.com/godot-jolt/godot-jolt/assets/97898580/cc316dcf-2336-478a-9eae-899ddcae6e55


 Here profiler.

**Reply by jrouwe** [reply]
> I know it not most helpful, but i think the bottleneck comes with memory use.

Very unlikely, Jolt mostly uses fixed buffer sizes that are configured on application startup time.

**Reply by Saul2022** [reply]
Alright, then it not responsible, though something weird  i noticed  is that godot physics has the same situation with half of the cubes, see  https://twitter.com/Saulcava1/status/1705897442825117766 , as jolt is way faster it produces more i understand that, but it weird having the same situation where fps suddenly turn from 50 or 24 to 5 or 3, and maybe it the current setup, but sadly decreasing iteration value, doesn’t help here. Or maybe it something related cache as you said. Or mihe.

**Reply by jrouwe** [reply]
On my machine the performance fall happens at around 8K bodies, the problem is that the bodies are in a big pile so the amount of contacts increases dramatically with pile size. If I move the 5x5 grid of vertical piles further apart (e.g. 30 m apart) then the performance with 8K bodies is still good on my machine.

**Reply by Saul2022** [reply]
True it happen's to me too, but  unity can handle them, could there be a possibility that it happen's because of a core physics  internal fro godot side?  Also in all the comparison, unity is always appearing  faster when the cubes fall, so maybe they do something to reduce calculations.

**Reply by Saul2022** [reply]
curious , this is now with  1.2k again with latest 0.8 , but this time is falls from far higher, it seem's gravity is affecting it. 

It seem's like gravity is maybe impacting the performance when falling, though when i added 2.4k and splitted them, one at right and one 1.2k left, performance was more or less  the same. Though it has also to do with  the step per ticks on  the more 3d section. 

edit : meanwhile in other engine gravity doesn´t affect framerate almost as much https://twitter.com/Saulcava1/status/1705896155249328206

edit: 3 also out of curiousity of your test tried it and  not good perf 

https://github.com/godot-jolt/godot-jolt/assets/97898580/09d9d7e5-7dd8-4b23-82df-8a27197ee407
